src/sentry/monitoring/queues.py

```diff
@@ -88,6 +108,60 @@ def get_queue_by_name(name):
             return queue
 
 
+queue_monitoring_cluster = redis.redis_clusters.get(CLUSTER_NAME)
+
+
+def is_queue_healthy(queue_name: str) -> bool:
+    # check if queue is healthy by pinging Redis
+    return not queue_monitoring_cluster.sismember(KEY_NAME, queue_name)
+
+
+def is_healthy(queue_size):
+    return queue_size < UNHEALTHY_QUEUE_SIZE_THRESHOLD
+
+
+def update_queue_stats(redis_cluster, unhealthy) -> None:
+    if not unhealthy:
+        return
+
+    with redis_cluster.pipeline(transaction=True) as pipeline:
+        pipeline.delete(KEY_NAME)
+        pipeline.sadd(KEY_NAME, *unhealthy)
+        # expire in 1min if we haven't checked in
+        pipeline.expire(KEY_NAME, 60)
+        pipeline.execute()
+
+
+def run_queue_stats_updater(redis_cluster: str) -> None:
+    # bonus point if we manage to use asyncio and launch all tasks at once
+    # in case we have many queues to check
+    cluster = redis.redis_clusters.get(redis_cluster)
+    from sentry.monitoring.queues import backend
+
+    if backend is None:
+        raise Exception("unknown broker type")
+
+    queue_history = {queue: 0 for queue in QUEUES}
+    while True:
+        new_sizes = backend.bulk_get_sizes(QUEUES)
+        for (queue, size) in new_sizes:
+            if is_healthy(size):
+                queue_history[queue] = 0
+            else:
+                queue_history[queue] += 1
```

Comment:
this is just a single counter, not a ring buffer. you only ever increment this, so once you cross the threshold, you will never go back.

Reply:
We reset it to 0 if the queue is healthy once.